### PR TITLE
[IMP] base, web: allow to group by properties

### PR DESCRIPTION
**Pull request by std-odoo**
Purpose
=======
Allow to group records by their property values,
like we can do with normal fields.

Technical
=========
Because there's no foreign key, when we group by a relational property
we need to check the existence of the ids in the query (and same for
selection and tags, because we might have value of a deleted
option / tag in database).

Task-3032464

### CODE REVIEW

File: odoo/models.py [hunk 1]
```diff
@@ -2419,6 +2460,126 @@ def _read_group_resolve_many2x_fields(self, data, fields):
...
+                            [(fullname, '>=', start), (fullname, '<', end)],
+                            domain,
+                        ]),
+                        '__range': {group_name: {'from': start, 'to': end}},
+                        group_name: babel.dates.format_date(
+                            d[group_name],
+                            format=display_formats[func],
+                            locale=get_lang(self.env).code
+                        ),
+                    })
```
Comment (thjo-odoo) [inline odoo/models.py:2579]: Is there a specific reason why you are using the update function instead of modifying directly the values, like you did for the other types ? Other than to update `__domain`, `__range` and the value associated to the groupby at the same time ?

**Review by thjo-odoo — COMMENTED** [review]
Hello @std-odoo,

Quite an impressive PR this is, nothing really jumped to me as wrong, so I just have some questions and some readability suggestions.

Cheers !

**Review by std-odoo — COMMENTED** [review]
@thjo-odoo Hi :) 

I did the changes and rebased after the read_group refactor

Cheers

File: addons/web/static/tests/search/search_bar_menu_tests.js [hunk 2]
```diff
@@ -1706,5 +1726,129 @@ QUnit.module("Search", (hooks) => {
...
+
+        QUnit.test("group by properties", async function (assert) {
+            assert.expect(10);
+
+            // variable set true when we fetch the definition
+            let definitionFetched = false;
+
+            async function mockRPC(route, { method, model, args, kwargs }) {
+                if (method === "web_search_read" && model === "parentModel") {
+                    definitionFetched = true;
```
Comment (Polymorphe57) [inline addons/web/static/tests/search/search_bar_menu_tests.js:1738]: use assert.step and assert.verifySteps instead of that variable

File: addons/web/static/tests/search/search_bar_menu_tests.js [hunk 3]
```diff
@@ -1706,5 +1726,129 @@ QUnit.module("Search", (hooks) => {
...
+                                        string: "My Integer",
+                                    },
+                                ],
+                            },
+                        ],
+                    };
+                }
+            }
+
+            const controlPanel = await makeWithSearch({
```
Comment (Polymorphe57) [inline addons/web/static/tests/search/search_bar_menu_tests.js:1780]: I think it should be called searchBar

File: addons/web/static/src/search/utils/misc.js [hunk 4]
```diff
@@ -22,5 +22,7 @@ export const GROUPABLE_TYPES = [
     "integer",
     "many2one",
     "many2many",
+    "properties",
```
[the remark this answers is not in the document]
Comment (Polymorphe57) [reply addons/web/static/src/search/utils/misc.js:25]: things to go wrong in pivot view with header dropdowns

File: addons/web/static/src/search/properties_group_by_item/properties_group_by_item.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    get isSingleParent() {
+        const uniqueNames = new Set(this.state.groupByItems.map((item) => item.definitionRecordId));
+        return uniqueNames.size < 2;
+    }
+
+    /**
+     * Dynamically load the definition, only when needed (if we open the dropdown).
+     */
+    async loadDefinition() {
```
Comment (Polymorphe57) [inline addons/web/static/src/search/properties_group_by_item/properties_group_by_item.js:37]: call it beforeOpen?

File: addons/web/static/src/search/properties_group_by_item/properties_group_by_item.xml [hunk 6]
```diff
@@ -0,0 +1,42 @@
+<templates xml:space="preserve">
+    <t t-name="web.PropertiesGroupByItem" owl="1">
+        <Dropdown togglerClass="'o_menu_item' + (isActive ? ' selected' : '')" beforeOpen.bind="loadDefinition">
+            <t t-set-slot="toggler">
+                Properties
```
Comment (Polymorphe57) [inline addons/web/static/src/search/properties_group_by_item/properties_group_by_item.xml:5]: no distinction between different field of type properties? Should it not be linked to the property String?

File: addons/web/static/src/search/search_model.js [hunk 7]
```diff
@@ -1141,6 +1141,89 @@ export class SearchModel extends EventBus {
...
+        for (const field of fields) {
+            if (field.type !== "properties") {
+                continue;
+            }
+
+            const definitionField = fields.find((f) => f.name === field.definition_record);
+            const definitionRecordModel = definitionField.relation;
+            const definitionRecordField = field.definition_record_field;
+
+            const result = await this._fetchPropertiesDefinition(
```
Comment (Polymorphe57) [inline addons/web/static/src/search/search_model.js:1164]: a domain should be added: no need for records with definitions = [].
Maybe also refactor the code of _fetchPropertiesDefinition using the new field service

File: addons/web/static/src/search/search_model.js [hunk 8]
```diff
@@ -1141,6 +1141,89 @@ export class SearchModel extends EventBus {
...
+                        !groupNames.includes(searchItem.name)
+                    ) {
+                        // we can not just remove the element from the list because index are used as id
+                        searchItem.type = "group_by_property_deleted";
+                    }
+                });
+
+                for (const definition of definitions) {
+                    const fullName = `${field.name}.${definition.name}`;
+                    this.searchViewFields[fullName] = {
```
Comment (Polymorphe57) [inline addons/web/static/src/search/search_model.js:1192]: could you add a comment explaining why you need that?

File: addons/web/static/src/search/search_model.js [hunk 9]
```diff
@@ -1141,6 +1141,89 @@ export class SearchModel extends EventBus {
...
+                );
+                Object.values(this.searchItems).forEach((searchItem) => {
+                    if (
+                        searchItem.isProperty &&
+                        searchItem.definitionRecordId === definitionRecordId &&
+                        ["groupBy", "dateGroupBy"].includes(searchItem.type) &&
+                        !groupNames.includes(searchItem.name)
+                    ) {
+                        // we can not just remove the element from the list because index are used as id
+                        searchItem.type = "group_by_property_deleted";
```
Comment (Polymorphe57) [inline addons/web/static/src/search/search_model.js:1186]: too smart. A comment would be most welcome

File: addons/web/static/tests/search/search_bar_menu_tests.js [hunk 10]
```diff
@@ -1706,5 +1726,129 @@ QUnit.module("Search", (hooks) => {
...
+                    "My Text (First Parent)",
+                    "My Partner (First Parent)",
+                    "My Datetime (First Parent)",
+                    "My Integer (Second Parent)",
+                ]
+            );
+
+            // open the datetime item
+            const datetimeItem = propertiesItems[2];
+            datetimeItem.dispatchEvent(new Event("mouseenter"));
```
[the remark this answers is not in the document]
Comment (Polymorphe57) [reply addons/web/static/tests/search/search_bar_menu_tests.js:1831]: you can use the helper mouseEnter from @web/../tests/helpers/utils

File: addons/web/static/src/views/relational_model.js [hunk 11]
```diff
@@ -3482,6 +3485,31 @@ export class RelationalModel extends Model {
         if (this.defaultOrderBy && !(params.orderBy && params.orderBy.length)) {
             rootParams.orderBy = this.defaultOrderBy;
         }
+
+        if (rootParams.groupBy && rootParams.groupBy.length) {
+            const groupByFieldName = rootParams.groupBy[0].split(":")[0];
```
Comment (Polymorphe57) [inline addons/web/static/src/views/relational_model.js:3490]: why just groupBy[0]?

Comment (Polymorphe57) [reply addons/web/static/src/views/relational_model.js:3490]: you should surely look in search model fields... You have fetched them already, no?

File: addons/web/static/src/views/relational_model.js [hunk 12]
```diff
@@ -3482,6 +3485,31 @@ export class RelationalModel extends Model {
...
+            if (groupByFieldName.includes(".") && !rootParams.fields[groupByFieldName]) {
```
Comment (Polymorphe57) [inline addons/web/static/src/views/relational_model.js:3491]: rewrite condition to make it clearer

File: addons/web/static/src/views/relational_model.js (hunk 11 above)
Comment (Polymorphe57) [reply addons/web/static/src/views/relational_model.js:3490]: check property keys of the fake fields....

**Review by std-odoo — COMMENTED** [review]
@Polymorphe57 Hi :) 

Thanks for the review, I did the changes and answered

**Review by ryv-odoo — COMMENTED** [review]
Hi @std-odoo , here is my first review (I didn't have the time to check the all diff yet) for the Python part.

**Review by std-odoo — COMMENTED** [review]
@ryv-odoo Hi :) 

Thanks for the review, I did the changes and answered :)

File: odoo/models.py [hunk 13]
```diff
@@ -2572,8 +2587,122 @@ def read_group(self, domain, fields, groupby, offset=0, limit=None, orderby=Fals
...
+            property_type = definition.get('type')
+
+            if property_type == 'selection':
+                options = definition.get('selection') or []
+                options = {option[0]: option[1] for option in options}
+                for d in data:
+                    if not d[fullname]:
+                        # can not do ('selection', '=', False) because we might have
+                        # option in database that does not exist anymore
+                        d['__domain'] = expression.AND([
```
[the remark this answers is not in the document]
Comment (ryv-odoo) [reply odoo/models.py:2619]: Whut ? `read_group` is not just for the list/kanban views. The rest of your code manages multiple groups, it is sad to be restricted here.

What about all the other view types (graph/pivot/gantt/...) ? (It crashes for graph by example :cry: )

[the remark this answers is not in the document]
Comment (std-odoo) [reply odoo/models.py:2619]: it crash for a different reason no ?

(they changed something and i didn't catch it after my many rebases...)

When I fixed the issue it work properly for many group, not sur to understand your point sorry


screenshot_spect

[the remark this answers is not in the document]
Comment (ryv-odoo) [reply odoo/models.py:2619]: The `__domain` is useful when you click on a group itself to show records in this group. 
In your case when `d[fullname]` is Falsy, it will rewrite `__domain` without taking in account `__domain` generates by a other groupby clause.

Here a example, groupby a property (many2one) and partner_id (normal field):
```JSON
[
  {
    "lead_properties.2b7505616fa5903a": false,
    "partner_id": [
      148,
      "Other"
    ],
    "__count": 1,
    "__domain": [
      "&",
      "|",
      [
        "lead_properties.2b7505616fa5903a",
        "=",
        false
      ],
      [
        "lead_properties.2b7505616fa5903a",
        "not in",
        []
      ],
      [
        "type",
        "=",
        "opportunity"
      ]
    ]
  },
  {
    "lead_properties.2b7505616fa5903a": false,
    "partner_id": [
      147,
      "STD"
    ],
    "__count": 1,
    "__domain": [
      "&",
      "|",
      [
        "lead_properties.2b7505616fa5903a",
        "=",
        false
      ],
      [
        "lead_properties.2b7505616fa5903a",
        "not in",
        []
      ],
      [
        "type",
        "=",
        "opportunity"
      ]
    ]
  }
]
```
You can see here that the `__domain` is wrong, it doesn't take in account the partner_id. Which leads to very weird behavior when you click on the group.

```Python
                       d['__domain'] = expression.AND([
                            expression.OR([
                                [(fullname, '=', False)],
                                [(fullname, 'not in', tuple(options))],
                            ]),
                            d['__domain'],
                        ])
```

Can you add some tests to cover multi-groupby?

**Review by ryv-odoo — COMMENTED** [review]
Another small batch

**Review by std-odoo — COMMENTED** [review]
@ryv-odoo Hi :) 

Thanks for this second batch, I did the changes and answered

Cheers

File: odoo/models.py [hunk 14]
```diff
@@ -2636,9 +2770,121 @@ def _inherits_join_calc(self, alias, fname, query):
...
+            return f'''
+                CASE
+                    WHEN jsonb_typeof({property_sql}) = 'string'
+                    THEN to_timestamp({property_sql}->>0, 'YYYY-MM-DD HH24:MI:SS')
+                    ELSE NULL
+                 END
+            '''
+
+        # if the key is not present in the dict, fallback to false instead of none
+        return f"COALESCE({property_sql}, 'false')"
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply odoo/models.py:2852]: it makes sens in JSON
```
>>> json.loads('false')
False
```

also when we group, the False column also means "no value" (e.g. if you have a boolean, we don't want 3 columns, "True", "False", "None")

File: odoo/fields.py [hunk 15]
```diff
@@ -3703,6 +3706,41 @@ def _dict_to_list(cls, values_dict, properties_definition):
...
+
+        Modify in place "values" to remove all occurrence of "properties.xxx"
+        and instead merge them into a single dict.
+
+        :param record: recordset on which we will write the values
+        :param values: the values to write, will be changed in place
+        """
+        all_properties_values = {}
+
+        for name, field_value in copy.copy(values).items():
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply odoo/fields.py]: as discussed at the coffee machine, it's better to not copy values (in some cases it can contains HTML, etc)

File: odoo/models.py [hunk 16]
```diff
@@ -2438,8 +2456,112 @@ def _read_group_format_result(self, rows_dict, lazy_groupby):
...
+
+        additional_domain = [(fullname, '=', row[group])]
+
+        property_type = definition.get('type')
+
+        if property_type == 'selection' and not row[fullname]:
+            # can not do ('selection', '=', False) because we might have
+            # option in database that does not exist anymore
+            options = definition.get('selection') or []
+            options = {option[0]: option[1] for option in options}
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply odoo/models.py]: woops, good catch

File: odoo/models.py [hunk 17]
```diff
@@ -2438,8 +2456,112 @@ def _read_group_format_result(self, rows_dict, lazy_groupby):
...
+            ])
+
+        elif property_type == 'many2one':
+            all_groups = tuple(
+                line[fullname] if isinstance(line[fullname], int)
+                else line[fullname][0]
+                for line in data if line[fullname]
+            )
+            if not row[fullname]:
+                # can not do ('many2one', '=', False) because we might have
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply odoo/models.py]: yes you are right :)

File: odoo/models.py [hunk 18]
```diff
@@ -2438,8 +2456,112 @@ def _read_group_format_result(self, rows_dict, lazy_groupby):
...
+            options = {option[0]: option[1] for option in options}
+            additional_domain = expression.OR([
+                [(fullname, '=', False)],
+                [(fullname, 'not in', tuple(options))],
+            ])
+
+        elif property_type == 'many2one':
+            all_groups = tuple(
+                line[fullname] if isinstance(line[fullname], int)
+                else line[fullname][0]
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply odoo/models.py]: it's because I modified the record in place, I can move it outside of the loop and I won't need that

File: odoo/models.py [hunk 19]
```diff
@@ -2438,8 +2456,112 @@ def _read_group_format_result(self, rows_dict, lazy_groupby):
...
+
+            elif (comodel := definition.get('comodel')) in self.env:
+                records = self.env[comodel].browse(all_groups)
+                name_gets = {record.id: (record.id, record.display_name) for record in records}
+                row[fullname] = name_gets.get(row[fullname])
+
+        elif property_type == 'tags':
+            tags = definition.get('tags') or []
+            tags = {tag[0]: tag for tag in tags}
+            # use the "in" for the domain and not "="
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply odoo/models.py]: It come from the search PR, we need the `IN` SQL operator, but for that we need to know the property type, which required a SQL query

Because of that we just used the `in` domain operator (so we know we should use the `IN` SQL operator)

**Review by std-odoo — COMMENTED** [review]
@ryv-odoo Hi :) 

Thanks for the review, I did the changes / answered 
(except one comment about the fallback in `read_progress_bar`, not sure we should handle it)

File: odoo/models.py [hunk 20]
```diff
@@ -84,23 +84,25 @@
...
     if not res_match:
         raise ValueError(
             f'Invalid aggregate/groupby specification {spec!r}.\n'
             '- Valid aggregate specification looks like "<field_name>:<agg>" example: "quantity:sum".\n'
             '- Valid groupby specification looks like "<no_datish_field_name>" or "<datish_field_name>:<granularity>" example: "date:month".'
         )
-    return res_match.groups()
+
+    groups = res_match.groups()
+    return groups[0], groups[2], groups[3]
```
Comment (tde-banana-odoo) [inline odoo/models.py:105]: Maybe update error message just above that contains some examples of expected input ?

File: odoo/models.py [hunk 21]
```diff
@@ -1865,9 +1867,11 @@ def _read_group(self, domain, groupby=(), aggregates=(), having=(), offset=0, li
         having_expression, having_params, fnames_used = self._read_group_having(having, query)
         fnames_to_flush.update(fnames_used)
 
-        orderby_terms, extra_groupby_terms, fnames_used = self._read_group_orderby(order, groupby, query)
+        orderby_terms, extra_groupby_terms, fnames_used = self._read_group_orderby(order, groupby_terms, query)
         fnames_to_flush.update(fnames_used)
 
+        groupby_terms = list(groupby_terms.values())
```
Comment (tde-banana-odoo) [inline odoo/models.py:1873]: It is not convenient to reuse a variable and change its type during a method ... makes code hard to follow. Can't you have another name for the dict-based "groupby_terms" ?

File: odoo/models.py [hunk 22]
```diff
@@ -1850,10 +1852,10 @@ def _read_group(self, domain, groupby=(), aggregates=(), having=(), offset=0, li
         query = self._search(domain)
 
         fnames_to_flush = OrderedSet()
-        groupby_terms = []  # [<SQL expression>,]
+        groupby_terms = {}  # {spec: <SQL expression>}
```
Comment (tde-banana-odoo) [inline odoo/models.py:1855]: groupby_field_sql_map ?

File: odoo/models.py [hunk 23]
```diff
@@ -2045,29 +2060,27 @@ def _read_group_orderby(self, order, groupby, query):
...
                 raise ValueError(f"Invalid order {order!r} for _read_group()")
-            order_field = order_match['field']
-            if order_match['property']:
-                raise ValueError(f"Invalid order {order!r} for _read_group(): properties are not supported")
+            order_term = order_match['term']
             order_direction = (order_match['direction'] or 'ASC').upper()
             order_nulls = (order_match['nulls'] or '').upper()
 
-            if order_field not in groupby:
+            if order_term not in groupby_terms:
```
Comment (tde-banana-odoo) [inline odoo/models.py:2067]: groupby_terms keys are the same as previous groupby items, isn't it ? Can't we keep the naming order_field ? order_fname ? Actually "term" is quite vague to me.

File: odoo/models.py [hunk 24]
```diff
@@ -2025,14 +2040,14 @@ def _read_group_having(self, having_domain, query):
         return expr, params, fnames_used
 
     @api.model
-    def _read_group_orderby(self, order, groupby, query):
+    def _read_group_orderby(self, order, groupby_terms, query):
```
Comment (tde-banana-odoo) [inline odoo/models.py:2043]: Maybe time to add a small docstring with input parameters, especially the type of 'groupby_terms' changes.

File: odoo/models.py [hunk 25]
```diff
@@ -1924,20 +1928,22 @@ def _read_group_select(self, aggregate_spec, query):
         if aggregate_spec == '__count':
             return 'COUNT(*)', []
 
-        fname, func = parse_read_group_spec(aggregate_spec)
+        fname, property_name, func = parse_read_group_spec(aggregate_spec)
+
+        fullname = f"{fname}.{property_name}" if property_name else fname
```
Comment (tde-banana-odoo) [inline odoo/models.py:1933]: sql_fname ? access_fname ? Don't like "fullname" as it hides a bit why we have two variables.

File: odoo/models.py [hunk 26]
```diff
@@ -1947,24 +1953,33 @@ def _read_group_groupby(self, groupby_spec, query):
...
+        if property_name:
+            check_property_field_value_name(property_name)
+            if field.type != "properties":
+                raise ValueError(f"Property set on a non properties field: {property_name!r}")
+
+            fullname = f"{fname}.{property_name}"
+        else:
+            fullname = fname
+
+        if granularity and field.type not in ('datetime', 'date', 'properties'):
```
Comment (tde-banana-odoo) [inline odoo/models.py:1971]: Shouldn't we check the property type here ?

File: odoo/models.py [hunk 27]
```diff
@@ -1947,24 +1953,33 @@ def _read_group_groupby(self, groupby_spec, query):
...
             raise ValueError(f"Granularity set on a no-datetime field: {groupby_spec!r}")
```
Comment (tde-banana-odoo) [inline odoo/models.py:1972]: Maybe error message could be more precise, like "non datetime field or property" ?

File: odoo/models.py [hunk 28]
```diff
@@ -2460,6 +2475,123 @@ def _read_group_format_result(self, rows_dict, lazy_groupby):
 
                 row['__domain'] = expression.AND([row['__domain'], additional_domain])
 
+    def _read_group_format_properties_result(self, rows_dict, group):
```
Comment (tde-banana-odoo) [inline odoo/models.py:2478]: ``_read_group_format_result_properties`` to keep the same base method name ?

File: odoo/models.py [hunk 29]
```diff
@@ -2460,6 +2475,123 @@ def _read_group_format_result(self, rows_dict, lazy_groupby):
...
+        """Modify the final read group properties result.
+
+        Replace the relational properties ids by a tuple with their display names,
+        replace the "raw" tags and selection values by a list containing their labels.
+        Adapt the domains for the Falsy group (we can't just keep (selection, =, False)
+        e.g. because some values in database might correspond to  option that have
+        been remove on the parent).
+        """
+        fullname, __, func = group.partition(':')
+        if '.' not in fullname:
```
Comment (tde-banana-odoo) [inline odoo/models.py:2488]: Could be done before partition.

File: odoo/models.py [hunk 30]
```diff
@@ -2460,6 +2475,123 @@ def _read_group_format_result(self, rows_dict, lazy_groupby):
...
+
+        elif property_type == 'many2many':
+            comodel = definition.get('comodel')
+            prefetch_ids = tuple(row[fullname] for row in rows_dict if row[fullname])
+            all_groups = tuple(row[fullname] for row in rows_dict if row[fullname])
+            for row in rows_dict:
+                if not row[fullname]:
+                    additional_domain = expression.OR([
+                        [(fullname, '=', False)],
+                        [(fullname, 'not in', group) for group in all_groups],
```
Comment (tde-banana-odoo) [inline odoo/models.py:2537]: I suppose the for-loop on the whole "not in" domain part (not on right part only) is due to the "not in" implementation ? Maybe expression.AND this sub-part ?

File: odoo/models.py [hunk 31]
```diff
@@ -2460,6 +2475,123 @@ def _read_group_format_result(self, rows_dict, lazy_groupby):
...
+                row['__domain'] = expression.AND([row['__domain'], additional_domain])
+
+        elif property_type == 'tags':
+            tags = definition.get('tags') or []
+            tags = {tag[0]: tag for tag in tags}
+            for row in rows_dict:
+                if not row[fullname]:
+                    additional_domain = expression.OR([
+                        [(fullname, '=', False)],
+                        [(fullname, 'not in', tag) for tag in tags],
```
Comment (tde-banana-odoo) [inline odoo/models.py:2553]: Same remark as for m2m

File: odoo/models.py [hunk 32]
```diff
@@ -2460,6 +2475,123 @@ def _read_group_format_result(self, rows_dict, lazy_groupby):
...
+        elif property_type in ('date', 'datetime'):
+            for row in rows_dict:
+                if not row[group]:
+                    row[group] = False
+                    row['__domain'] = expression.AND([row['__domain'], [(fullname, '=', False)]])
+                    row['__range'] = {}
+                    continue
+
+                # Date / Datetime are not JSONifiable, so they are stored as raw text
+                db_format = '%Y-%m-%d' if property_type == 'date' else '%Y-%m-%d %H:%M:%S'
```
Comment (tde-banana-odoo) [inline odoo/models.py:2571]: Shouldn't we use DEFAULT_SERVER_DATE_FORMAT and DEFAULT_SERVER_DATETIME_FORMAT ?

File: odoo/models.py [hunk 33]
```diff
@@ -2460,6 +2475,123 @@ def _read_group_format_result(self, rows_dict, lazy_groupby):
...
+
+                if func != 'week':
```
Comment (tde-banana-odoo) [inline odoo/models.py:2573]: Just for readability, I would do "if func == 'week' - else", as positive conditions are always easier to follow when browsing code.

File: odoo/models.py [hunk 34]
```diff
@@ -2508,10 +2640,12 @@ def read_group(self, domain, fields, groupby, offset=0, limit=None, orderby=Fals
         # - Modify the order to be compatible with the _read_group specification
         annoted_groupby = {}  # Key as the name in the result, value as the explicit groupby specification
         for group_spec in lazy_groupby:
-            field_name, granularity = parse_read_group_spec(group_spec)
+            field_name, property_name, granularity = parse_read_group_spec(group_spec)
             if field_name not in self._fields:
                 raise ValueError(f"Invalid field {field_name!r} on model {self._name!r}")
             field = self._fields[field_name]
+            if property_name and field.type != 'properties':
+                raise ValueError("Property name set on a non-properties fields.")
```
Comment (tde-banana-odoo) [inline odoo/models.py:2648]: Maybe "Property name {property_name!r} ha to be used on a property field" ?

File: odoo/tools/query.py [hunk 35]
```diff
@@ -57,6 +57,7 @@ def __init__(self, cr, alias, table=None):
 
         # joins {alias: (kind, table, condition, condition_params)}
         self._joins = {}
+        self._raw_joins = {}  # {query: params}
```
Comment (tde-banana-odoo) [inline odoo/tools/query.py:60]: Why not a list of 2-items tuples maybe ? Do we need the specific key-based behavior ?

**Review by tde-banana-odoo — COMMENTED** [review]
Hello @std-odoo ,

Sorry for the delay, finally took some time to look at it :) glad to see this feature, it is effectively interesting to be able to group on properties !

Added some comments while reading the code, mainly some beautification suggestions and some questions. Thanks for your work !

**Review by std-odoo — COMMENTED** [review]
@tde-banana-odoo Hi :) 

I did the changes and answered :)

File: addons/web/static/src/model/relational_model/relational_model.js [hunk 36]
```diff
@@ -182,6 +182,10 @@ export class RelationalModel extends Model {
      */
     async load(params = {}) {
         const config = this._getNextConfig(this.config, params);
+        const groupByFieldName = config?.groupBy?.[0]?.split(":")[0] || "";
```
Comment (FrancoisGe) [inline addons/web/static/src/model/relational_model/relational_model.js:185]: config? => config

**Review by std-odoo — COMMENTED** [review]
@aab-odoo @FrancoisGe Hi :) 

I applied the comments we discussed IRL, I added the tests for the pivot view / drag & drop in the last commit, because I needed to make some changes to `MockServer` / `Domain` (because it simulates the server `read_group` / `search_read`)

Cheers

File: addons/web/static/src/model/relational_model/record.js [hunk 37]
```diff
@@ -942,10 +942,22 @@ export class Record extends DataPoint {
...
-                const propertyFieldName = field.relatedPropertyField.fieldName;
-                changes[propertyFieldName] = this.data[propertyFieldName].map((property) =>
-                    property.name === field.propertyName ? { ...property, value } : property
+                const [propertyFieldName, propertyName] = field.name.split(".");
+                const propertiesData = this.data[propertyFieldName] || [];
+                if (!propertiesData.find((property) => property.name === propertyName)) {
+                    // try to change the value of a properties that has a different parent
+                    this.model.notification.add(
+                        _t(
+                            "This record belongs to a different parent so you can not change this property."
```
Comment (aab-odoo) [inline addons/web/static/src/model/relational_model/record.js:951]: I'm wondering if a random user will understand this message

File: addons/web/static/src/model/relational_model/relational_model.js [hunk 38]
```diff
@@ -180,6 +181,18 @@ export class RelationalModel extends Model {
      */
     async load(params = {}) {
         const config = this._getNextConfig(this.config, params);
+        const groupByFieldName = config.groupBy?.[0]?.split(":")[0] || "";
```
Comment (aab-odoo) [inline addons/web/static/src/model/relational_model/relational_model.js:184]: maybe move this in _loadGroupedList ? There, you know what you have a groupBy, and it's not empty

File: addons/web/static/src/model/relational_model/relational_model.js [hunk 39]
```diff
@@ -223,6 +236,30 @@ export class RelationalModel extends Model {
...
+        if (!result) {
+            // the property might have been removed
+            config.groupBy = null;
+        } else {
+            result.propertyName = result.name;
+            result.name = propertyFullName; // "xxxxx" -> "property.xxxxx"
+            // needed for _applyChanges
+            result.relatedPropertyField = { fieldName: propertyFullName.split(".")[0] };
+            result.relation = result.comodel; // match name on field
+            config.fields[propertyFullName] = result;
```
Comment (aab-odoo) [inline addons/web/static/src/model/relational_model/relational_model.js:259]: maybe this rpc should return the field directly, instead of something else from which you build a field in js.

File: addons/web/static/src/model/relational_model/relational_model.js [hunk 40]
```diff
@@ -223,6 +236,30 @@ export class RelationalModel extends Model {
...
+        // dynamically load the property and add the definition in the fields attribute
+        const result = await this.orm.call(
+            config.resModel,
+            "get_property_definition",
+            [propertyFullName],
+            { context: config.context }
+        );
+        if (!result) {
+            // the property might have been removed
+            config.groupBy = null;
```
Comment (aab-odoo) [inline addons/web/static/src/model/relational_model/relational_model.js:252]: Note that this code also applies in list view, where you can have several level on groupbys, in this case, you remove all the groupbys.

Also, you can groupby several fields, and if there's a property among them, but not as first groupby, your code won't be executed.

File: addons/web/static/tests/helpers/mock_server.js [hunk 41]
```diff
@@ -2178,6 +2191,18 @@ export class MockServer {
...
                         criterion = [criterion[0], "in", [criterion[2]]];
                     }
+                } else if (field.relatedPropertyField) {
+                    // Ugly hack to not change @Domain, we transform the list into a dict
+                    const fielName = field.relatedPropertyField.name;
+                    records = records.map((record) => {
+                        const newRecord = { ...record };
+                        const properties = record[fielName] || [];
+                        newRecord[`__${fielName}`] = Object.fromEntries(
+                            new Map(properties.map((property) => [property.name, property.value]))
```
Comment (Polymorphe57) [inline addons/web/static/tests/helpers/mock_server.js:2201]: new Mapis not necessary I think

File: addons/web/static/tests/helpers/mock_server.js [hunk 42]
```diff
@@ -2178,6 +2191,18 @@ export class MockServer {
                     } else {
                         criterion = [criterion[0], "in", [criterion[2]]];
                     }
+                } else if (field.relatedPropertyField) {
+                    // Ugly hack to not change @Domain, we transform the list into a dict
```
Comment (Polymorphe57) [inline addons/web/static/tests/helpers/mock_server.js:2195]: an object

dict is not the usual terminology in js

File: addons/web/static/tests/views/pivot_view_tests.js [hunk 43]
```diff
@@ -125,6 +125,27 @@ QUnit.module("Views", (hooks) => {
...
+                            type: "many2one",
+                            relation: "partner",
+                            name: "parent_id",
+                        },
+                        properties_definition: {
+                            string: "Properties",
+                            type: "properties_definition",
+                        },
+                        "properties.my_char": {
+                            // needed for MockServer to work, see @mockReadGroup
```
Comment (Polymorphe57) [inline addons/web/static/tests/views/pivot_view_tests.js:146]: I think it would be better to define yourself the return of webReadGroup, webSearchRead in your tests instead of doing this kind of things. You have mockRPC for that. The code in mock_server is very difficult to follow and make it rely on this kind of tricks is really weird

File: addons/web/static/src/model/relational_model/relational_model.js (hunk 40 above)
Comment (std-odoo) [reply addons/web/static/src/model/relational_model/relational_model.js:252]: > Note that this code also applies in list view, where you can have several level on groupbys, in this case, you remove all the groupbys.

This is ok I think, it's already a edge case where you have a default group by, and you remove the property (the user will have to re-do the group by anyway)

File: addons/web/static/src/model/relational_model/relational_model.js (hunk 38 above)
Comment (std-odoo) [reply addons/web/static/src/model/relational_model/relational_model.js:184]: better indeed, done

File: addons/web/static/tests/helpers/mock_server.js (hunk 41 above)
Comment (std-odoo) [reply addons/web/static/tests/helpers/mock_server.js:2201]: good catch

File: odoo/models.py [hunk 44]
```diff
@@ -2650,9 +2793,125 @@ def _inherits_join_calc(self, alias, fname, query):
...
+        target_model = self.env[self._fields[field.definition_record].comodel_name]
+        query = """
+            SELECT definition
+              FROM {table},
+                   jsonb_array_elements({field}) definition
+             WHERE {field} IS NOT NULL
+               AND definition->>'name' = %s
+             LIMIT 1
+        """.format(table=target_model._table, field=field.definition_record_field)
+        self.env.cr.execute(query, [property_name])
```
Comment (tde-banana-odoo) [inline odoo/models.py:2911]: This will have to be written using sql.Identifiers and named parameters I think, like
```suggestion
        query = sql.SQL("""
            SELECT definition
              FROM {table},
                   jsonb_array_elements({field}) definition
             WHERE {field} IS NOT NULL
               AND definition->>'name' = %{property_name}s
             LIMIT 1
        """.format(table=sql.Indentifier(target_model._table), field=sql.Indentifier(field.definition_record_field))
        self.env.cr.execute(query, {'property_name': property_name})
```

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge

Thanks for your work and thanks to the various framework teams reviewers :)

File: addons/web/static/src/model/relational_model/relational_model.js (hunk 40 above)
Comment (aab-odoo) [reply addons/web/static/src/model/relational_model/relational_model.js:203]: What about
 
> Also, you can groupby several fields, and if there's a property among them, but not as first groupby, your code won't be executed.

?

Comment (std-odoo) [reply addons/web/static/src/model/relational_model/relational_model.js:203]: it work, if you save a group by with a property as second, the property definition will be fetched when you unfold the list

image
